interp/iostream.go

```diff
@@ -0,0 +1,203 @@
+package interp
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127
+)
+
+func cmdExitCode(cmdError error, ioerrs... error) (exitCode int, err error) {
+	var ee *exec.ExitError
+	var ok bool
+
+	if cmdError == nil {
+		// exitCode = 0.0
+		goto CheckIOErrors
+	}
+	if ee, ok = cmdError.(*exec.ExitError); !ok {
+		exitCode = -1.0
+		return
+	}
+	if exitCode, err = exitErrorCode(ee); err != nil {
+		return
+	}
+
+CheckIOErrors:
+	for _, err = range ioerrs {
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+// Copy the behavior of [gm]awk for pipe exit status. n?awk returns the raw
+// wait(3) return value but this is hard to use without a bit shift operator!
+func exitErrorCode(ee *exec.ExitError) (exitCode int, err error) {
+	stat, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Not all platforms support WaitStatus?
+		return -1.0, err
+	}
+	switch {
+	case stat.Exited():
+		exitCode = stat.ExitStatus()
+	case stat.Signaled():
+		exitCode = 256 + int(stat.Signal())
+		if stat.CoreDump() {
+			exitCode += 256
+		}
+	default:
+		exitCode = -1
+		err = ee
+	}
+	return
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() int
+}
+
+type outputStream interface {
+	io.WriteCloser
+	Flush() error
+	ExitCode() int
+}
+
+type outFileStream struct {
+	*bufio.Writer
+	io.Closer
+	exitCode int
+}
+
+func outputFileStream(wc io.WriteCloser, size int) outputStream {
+	b := bufio.NewWriterSize(wc, size)
+	return &outFileStream{b, wc, notClosedExitCode}
+}
+
+func (s *outFileStream) Close() (err error) {
+	err1 := s.Writer.Flush()
+	err2 := s.Closer.Close()
+	switch {
+	case err1 != nil:
+		s.exitCode, err = -1, err1
+	case err2 != nil:
+		s.exitCode, err = -1, err2
+	default:
+		s.exitCode = 0
+	}
+	return
+}
+
+func (s *outFileStream) ExitCode() int {
+	return s.exitCode
+}
+
+type outCmdStream struct {
+	io.Closer
+	*bufio.Writer
+	cmd *exec.Cmd
+	exitCode int
+}
+
+func outputCmdStream(cmd *exec.Cmd) (outputStream, error) {
+	w, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, newError("error connecting to stdin pipe: %v", err)
+	}
+	err = cmd.Start()
+	if err != nil {
+		w.Close()
+		return nil, err
+	}
+	out := &outCmdStream{w, bufio.NewWriterSize(w, outputBufSize), cmd, notClosedExitCode}
+	return out, nil
+}
+
+func (s *outCmdStream) Close() (err error) {
+	err1 := s.Writer.Flush()
+	err2 := s.Closer.Close()
+	s.exitCode, err = cmdExitCode(s.cmd.Wait(), err1, err2)
+	return
+}
+
+func (s *outCmdStream) ExitCode() int {
+	return s.exitCode
+}
+
+type outNullStream struct {
+	io.Writer
+}
+
+func outputNullStream() outputStream {
+	return &outNullStream{io.Discard}
+}
+
+func (s *outNullStream) Flush() error {
+	return nil
+}
+
+func (s *outNullStream) Close() error {
+	return nil
+}
+
+func (s *outNullStream) ExitCode() int {
+	return -1
+}
+
+type inFileStream struct {
+	io.ReadCloser
```

Comment:
I don't feel that would be an improvement. Creating a separate `closer` field in `outCmdStream` made sense because we were not using embedding to pass through the `Close()` method to the `io.Closer`. The `io.Closer` of course only offers the `Close()` method so it was pointless to embed it. That is not the case here.